src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -1324,13 +1324,15 @@ void InterpreterSelectQuery::executeImpl(QueryPlan & query_plan, const BlockInpu
                     }
             }
 
+            bool apply_limit = options.to_stage != QueryProcessingStage::WithMergeableStateAfterAggregation;
+            bool apply_offset = options.to_stage != QueryProcessingStage::WithMergeableStateAfterAggregationAndLimit;
```

Comment:
Why offset is not applied at `WithMergeableStateAfterAggregationAndLimit` stage?

Reply:
If the offset will be applied for `WithMergeableStateAfterAggregationAndLimit` (the stage up to which the query is processed on the remote shard) then the initiator will not receive correct rows to apply OFFSET with ORDER BY (since each shard will already apply offset and the global order will be incorrect).

I will add a test for this.